### ISSUE TITLE
Physical keyboard produces text from non-alphanumerical keys

### ISSUE THREAD

**Issue by clickworkorange** [kivy/kivy#8152]
**Software Versions**
* Python: 3.7.3
* OS: Raspbian Buster 
* Window provider: egl_rpi
* Kivy: 2.1.0
* Kivy installation method: built from source

**Describe the bug**
Pressing for example the caps lock key on a physical keyboard while a TextInput has focus prints the string representation of the key, in this case "capslock". The same is true for the function keys (e.g. pressing F1 prints "f1" in the TextBox). 

**Expected behavior**
Non-alphanumeric keys should remain "silent". 

**To Reproduce**
Run the example below with a physical keyboard connected and configured in .kivy/config.ini Try pressing some non-alphanumeric keys. 

**Code and Logs and screenshots**
```python
#!/usr/bin/env python3

from kivy.core.window import Window
from kivy.app import App
from kivy.lang import Builder
from kivy.uix.textinput import TextInput
from kivy.uix.boxlayout import BoxLayout

kv = '''
TextInput:
    focus: True
    font_size: dp(64)
'''

class MyApp(App):
  def __init__(self, **kwargs):
    super(MyApp, self).__init__(**kwargs)
    self.keyboard = Window.request_keyboard(self.on_keyboard_close, self)
    self.keyboard.bind(on_key_down=self.on_key_down)

  def on_key_down(self, keyboard, keycode, text, modifiers):
    print(keycode)

  def on_keyboard_close(self):
    return True

  def build(self):
    return Builder.load_string(kv)

if __name__ == '__main__':
  app = MyApp()
  try:
    app.run()
  except KeyboardInterrupt:
    app.stop()
```

**Additional context**
This is on a Raspberry Pi with no X11, so Kivy uses egl_rpi for window management. Keyboard input is still through XKB though (I think?).

**Comment by clickworkorange** [comment]
Some further testing reveals that [delete], [backspace], [home], [end], [PgUp], [PgDn] and the arrow keys produce no output, while [Caps Lock], [Num Lock], [F1] - [F12], [Insert], [Print Screen], [Scroll Lock] and [Pause] do.

**Comment by clickworkorange** [comment]
Perhaps related, Kivy also does not respect the keyboard layout configured with `dpkg-reconfigure keyboard-configuration`, `localectl set-keymap` and `localectl set-x11-keymap`. This works on the console, but I Kivy the layout is always en-US.

**Comment by misl6** [comment]
Hi @clickworkorange !

> This is on a Raspberry Pi with no X11, so Kivy uses egl_rpi for window management. Keyboard input is through SDL2, I think?

Can you please post the startup (debug) log?

**Comment by clickworkorange** [comment]
@misl6 Sure, is this what you mean? 

````
$ python3 my_app.py  -c kivy:log_level:debug
[INFO   ] [Logger      ] Record log in /home/myself/.kivy/logs/kivy_23-02-26_56.txt
[INFO   ] [Kivy        ] v2.1.0
[INFO   ] [Kivy        ] Installed at "/usr/local/lib/python3.7/dist-packages/kivy/__init__.py"
[INFO   ] [Python      ] v3.7.3 (default, Oct 31 2022, 14:04:00) 
[GCC 8.3.0]
[INFO   ] [Python      ] Interpreter at "/usr/bin/python3"
[INFO   ] [Logger      ] Purge log fired. Processing...
[INFO   ] [Logger      ] Purge finished!
[INFO   ] [Factory     ] 189 symbols loaded
[DEBUG  ] [Cache       ] register <kv.resourcefind> with limit=None, timeout=60
[DEBUG  ] [Cache       ] register <kv.lang> with limit=None, timeout=None
[DEBUG  ] [Cache       ] register <kv.image> with limit=None, timeout=60
[DEBUG  ] [Cache       ] register <kv.atlas> with limit=None, timeout=None
[INFO   ] [Image       ] Providers: img_tex, img_dds, img_sdl2 (img_pil, img_ffpyplayer ignored)
[DEBUG  ] [Cache       ] register <kv.texture> with limit=1000, timeout=60
[DEBUG  ] [Cache       ] register <kv.shader> with limit=1000, timeout=3600
[INFO   ] [Window      ] Provider: egl_rpi
[DEBUG  ] [Window      ] Actual display size: 1920x1080
[INFO   ] [GL          ] Using the "OpenGL ES 2" graphics system
[DEBUG  ] [GL          ] glShaderBinary is not available
[INFO   ] [GL          ] Backend used <gl>
[INFO   ] [GL          ] OpenGL version <b'OpenGL ES 2.0'>
[INFO   ] [GL          ] OpenGL vendor <b'Broadcom'>
[INFO   ] [GL          ] OpenGL renderer <b'VideoCore IV HW'>
[INFO   ] [GL          ] OpenGL parsed version: 2, 0
[INFO   ] [GL          ] Shading version <b'OpenGL ES GLSL ES 1.00'>
[INFO   ] [GL          ] Texture max size <2048>
[INFO   ] [GL          ] Texture max units <8>
[DEBUG  ] [Shader      ] Fragment compiled successfully
[DEBUG  ] [Shader      ] Vertex compiled successfully
[DEBUG  ] [ImageSDL2   ] Load </usr/local/lib/python3.7/dist-packages/kivy/data/glsl/default.png>
[INFO   ] [Window      ] virtual keyboard not allowed, single mode, not docked
[INFO   ] [Text        ] Provider: sdl2
[DEBUG  ] [Cache       ] register <textinput.label> with limit=None, timeout=60.0
[DEBUG  ] [Cache       ] register <textinput.width> with limit=None, timeout=60.0
[DEBUG  ] [App         ] Loading kv <./my_app.kv>
[DEBUG  ] [Resource    ] add </usr/share/fonts> in path list
[DEBUG  ] [Resource    ] add </usr/share/fonts/truetype> in path list
[DEBUG  ] [Resource    ] add </usr/share/fonts/truetype/dejavu> in path list
[DEBUG  ] [Resource    ] add </usr/local/share/fonts> in path list
[DEBUG  ] [Resource    ] add </usr/local/lib/python3.7/dist-packages/kivy/data/fonts> in path list
[DEBUG  ] [Atlas       ] Load </usr/local/lib/python3.7/dist-packages/kivy/data/images/defaulttheme.atlas>
[DEBUG  ] [Atlas       ] Need to load 1 images
[DEBUG  ] [Atlas       ] Load </usr/local/lib/python3.7/dist-packages/kivy/data/images/defaulttheme-0.png>
[DEBUG  ] [ImageSDL2   ] Load </usr/local/lib/python3.7/dist-packages/kivy/data/images/defaulttheme-0.png>
[DEBUG  ] [ImageSDL2   ] Load </home/myself/progress_track.png>
[DEBUG  ] [ImageSDL2   ] Load </home/myself/progress_bar.png>
[INFO   ] [GL          ] NPOT texture support is available
[INFO   ] [Clipboard   ] Provider: xclip
[INFO   ] [CutBuffer   ] cut buffer support enabled
[DEBUG  ] [Base        ] Create provider from hidinput,/dev/input/event0
[INFO   ] [HIDInput    ] Read event from </dev/input/event0>
[INFO   ] [Base        ] Start application main loop
[INFO   ] [HIDMotionEvent] using <Chicony HP Elite USB Keyboard>
````
Please let me know if there's any other information that may be helpful - I'm very eager to get this is issue fixed!

**Comment by misl6** [comment]
Ok, and can you share the content of `config.ini` file located at `/home/myself/.kivy/config.ini` ?

**Comment by clickworkorange** [comment]
@misl6 Here you go: 

````
[kivy]
keyboard_repeat_delay = 300
keyboard_repeat_rate = 30
log_dir = logs
log_enable = 1
log_level = info
log_name = kivy_%y-%m-%d_%_.txt
window_icon = 
keyboard_mode = system
keyboard_layout = qwerty
desktop = 0
exit_on_escape = 0
pause_on_minimize = 0
kivy_clock = default
default_font = ['Roboto', 'data/fonts/Roboto-Regular.ttf', 'data/fonts/Roboto-Italic.ttf', 'data/fonts/Roboto-Bold.ttf', 'data/fonts/Roboto-BoldItalic.ttf']
log_maxfiles = 100
window_shape = data/images/defaultshape.png
config_version = 24

[graphics]
display = -1
fullscreen = auto
width = 1920
height = 1080
top = 0
left = 0
maxfps = 30
multisamples = 2
position = auto
rotation = 0
show_cursor = 1
resizable = 0
borderless = 1
window_state = visible
minimum_width = 0
minimum_height = 0
min_state_time = .035
allow_screensaver = 0
shaped = 0
vsync = 
verify_gl_main_thread = 1
custom_titlebar = 0
custom_titlebar_border = 5

[input]
#mouse = mouse
#hid_%(name)s = probesysfs,provider=hidinput
keyboard = hidinput,/dev/input/event0
#mouse = hidinput,/dev/input/event3

[postproc]
double_tap_distance = 20
double_tap_time = 250
ignore = []
jitter_distance = 0
jitter_ignore_devices = mouse,mactouch,
retain_distance = 50
retain_time = 0
triple_tap_distance = 20
triple_tap_time = 375

[widgets]
scroll_timeout = 250
scroll_distance = 20
scroll_friction = 1.
scroll_stoptime = 300
scroll_moves = 5

[modules]
# cursor = 1
# inspector = 1
# showborder = 1

[network]
useragent = curl
````

**Comment by misl6** [comment]
What happens if you don't specifically configure the keyboard via `config.ini` and let Kivy decide ?

**Comment by clickworkorange** [comment]
@misl6 You mean by using the `hid_%(name)s = probesysfs,provider=hidinput` line in config.ini instead of `keyboard = hidinput,/dev/input/event0`? Exactly the same behaviour I'm afraid, though the debug console output on start has some additional bits in it: 

````
[DEBUG  ] [Base        ] Create provider from probesysfs,provider=hidinput
[DEBUG  ] [ProbeSysfs  ] using probesysfs!
[DEBUG  ] [ProbeSysfs  ] found device: Microsoft  Microsoft Basic Optical Mouse v2.0 at /dev/input/event3
[INFO   ] [ProbeSysfs  ] device match: /dev/input/event3
[INFO   ] [HIDInput    ] Read event from </dev/input/event3>
[DEBUG  ] [ProbeSysfs  ] found device: Chicony HP Elite USB Keyboard System Control at /dev/input/event1
[INFO   ] [ProbeSysfs  ] device match: /dev/input/event1
[INFO   ] [HIDInput    ] Read event from </dev/input/event1>
[DEBUG  ] [ProbeSysfs  ] found device: Chicony HP Elite USB Keyboard Consumer Control at /dev/input/event2
[INFO   ] [ProbeSysfs  ] device match: /dev/input/event2
[INFO   ] [HIDInput    ] Read event from </dev/input/event2>
[DEBUG  ] [ProbeSysfs  ] found device: Chicony HP Elite USB Keyboard at /dev/input/event0
[INFO   ] [ProbeSysfs  ] device match: /dev/input/event0
[INFO   ] [HIDInput    ] Read event from </dev/input/event0>
[INFO   ] [Base        ] Start application main loop
[INFO   ] [HIDMotionEvent] using <Microsoft  Microsoft Basic Optical Mouse v2.0 >
[INFO   ] [HIDMotionEvent] using <Chicony HP Elite USB Keyboard System Control>
[INFO   ] [HIDMotionEvent] using <Chicony HP Elite USB Keyboard Consumer Control>
[INFO   ] [HIDMotionEvent] using <Chicony HP Elite USB Keyboard>
````
Most of which is not needed for my application. I also lose the ability to disable the mouse; it gets picked up regardless of the `mouse = ` setting.

**Comment by clickworkorange** [comment]
I might add that I've just updated SDL2 to 2.26.3, which is the most recent release, partly in the hope that this might fix the issue. It did not. But I'm actually not sure if keyboard input in Kivy comes courtesy of SDL2, or if it's from somewhere else. I did notice I had to install `xkb` and `xclip`, despite not using X11, so perhaps `xkb` is handling it?

**Comment by misl6** [comment]
> @misl6 You mean by using the `hid_%(name)s = probesysfs,provider=hidinput` line in config.ini instead of `keyboard = hidinput,/dev/input/event0`? Exactly the same behaviour I'm afraid, though the debug console output on start has some additional bits in it:
> 
> ```
> [DEBUG  ] [Base        ] Create provider from probesysfs,provider=hidinput
> [DEBUG  ] [ProbeSysfs  ] using probesysfs!
> [DEBUG  ] [ProbeSysfs  ] found device: Microsoft  Microsoft Basic Optical Mouse v2.0 at /dev/input/event3
> [INFO   ] [ProbeSysfs  ] device match: /dev/input/event3
> [INFO   ] [HIDInput    ] Read event from </dev/input/event3>
> [DEBUG  ] [ProbeSysfs  ] found device: Chicony HP Elite USB Keyboard System Control at /dev/input/event1
> [INFO   ] [ProbeSysfs  ] device match: /dev/input/event1
> [INFO   ] [HIDInput    ] Read event from </dev/input/event1>
> [DEBUG  ] [ProbeSysfs  ] found device: Chicony HP Elite USB Keyboard Consumer Control at /dev/input/event2
> [INFO   ] [ProbeSysfs  ] device match: /dev/input/event2
> [INFO   ] [HIDInput    ] Read event from </dev/input/event2>
> [DEBUG  ] [ProbeSysfs  ] found device: Chicony HP Elite USB Keyboard at /dev/input/event0
> [INFO   ] [ProbeSysfs  ] device match: /dev/input/event0
> [INFO   ] [HIDInput    ] Read event from </dev/input/event0>
> [INFO   ] [Base        ] Start application main loop
> [INFO   ] [HIDMotionEvent] using <Microsoft  Microsoft Basic Optical Mouse v2.0 >
> [INFO   ] [HIDMotionEvent] using <Chicony HP Elite USB Keyboard System Control>
> [INFO   ] [HIDMotionEvent] using <Chicony HP Elite USB Keyboard Consumer Control>
> [INFO   ] [HIDMotionEvent] using <Chicony HP Elite USB Keyboard>
> ```
> 
> Most of which is not needed for my application. I also lose the ability to disable the mouse; it gets picked up regardless of the `mouse = ` setting. All minor niggles which I could happily live with if it fixed the problem, which unfortunately it doesn't.

How about using `hid_%(name)s = probesysfs` instead of `hid_%(name)s = probesysfs,provider=hidinput` ?
(Unfortunately ATM, I do not have an RPi to debug on)

**Comment by clickworkorange** [comment]
@misl6 That has the odd effect that although the keyboard (and mouse) is detected no keyboard events are fired when pressing the keys. In other words the keyboard stops working altogether. Startup debug log: 

````
$ python3 my_app.py  -c kivy:log_level:debug
[INFO   ] [Logger      ] Record log in /home/myself/.kivy/logs/kivy_23-03-06_8.txt
[INFO   ] [Kivy        ] v2.1.0
[INFO   ] [Kivy        ] Installed at "/usr/local/lib/python3.7/dist-packages/kivy/__init__.py"
[INFO   ] [Python      ] v3.7.3 (default, Oct 31 2022, 14:04:00) 
[GCC 8.3.0]
[INFO   ] [Python      ] Interpreter at "/usr/bin/python3"
[INFO   ] [Logger      ] Purge log fired. Processing...
[INFO   ] [Logger      ] Purge finished!
[INFO   ] [Factory     ] 189 symbols loaded
[DEBUG  ] [Cache       ] register <kv.resourcefind> with limit=None, timeout=60
[DEBUG  ] [Cache       ] register <kv.lang> with limit=None, timeout=None
[DEBUG  ] [Cache       ] register <kv.image> with limit=None, timeout=60
[DEBUG  ] [Cache       ] register <kv.atlas> with limit=None, timeout=None
[INFO   ] [Image       ] Providers: img_tex, img_dds, img_sdl2 (img_pil, img_ffpyplayer ignored)
[DEBUG  ] [Cache       ] register <kv.texture> with limit=1000, timeout=60
[DEBUG  ] [Cache       ] register <kv.shader> with limit=1000, timeout=3600
[INFO   ] [Window      ] Provider: egl_rpi
[DEBUG  ] [Window      ] Actual display size: 1920x1080
[INFO   ] [GL          ] Using the "OpenGL ES 2" graphics system
[DEBUG  ] [GL          ] glShaderBinary is not available
[INFO   ] [GL          ] Backend used <gl>
[INFO   ] [GL          ] OpenGL version <b'OpenGL ES 2.0'>
[INFO   ] [GL          ] OpenGL vendor <b'Broadcom'>
[INFO   ] [GL          ] OpenGL renderer <b'VideoCore IV HW'>
[INFO   ] [GL          ] OpenGL parsed version: 2, 0
[INFO   ] [GL          ] Shading version <b'OpenGL ES GLSL ES 1.00'>
[INFO   ] [GL          ] Texture max size <2048>
[INFO   ] [GL          ] Texture max units <8>
[DEBUG  ] [Shader      ] Fragment compiled successfully
[DEBUG  ] [Shader      ] Vertex compiled successfully
[DEBUG  ] [ImageSDL2   ] Load </usr/local/lib/python3.7/dist-packages/kivy/data/glsl/default.png>
[INFO   ] [Window      ] virtual keyboard not allowed, single mode, not docked
[INFO   ] [Text        ] Provider: sdl2
[DEBUG  ] [Cache       ] register <textinput.label> with limit=None, timeout=60.0
[DEBUG  ] [Cache       ] register <textinput.width> with limit=None, timeout=60.0
[DEBUG  ] [Audio       ] register SoundSDL2
[DEBUG  ] [Audio       ] register MusicSDL2
[INFO   ] [Audio       ] Providers: audio_sdl2 (audio_ffpyplayer ignored)
[DEBUG  ] [App         ] Loading kv <./my_app.kv>
[DEBUG  ] [Resource    ] add </usr/share/fonts> in path list
[DEBUG  ] [Resource    ] add </usr/share/fonts/truetype> in path list
[DEBUG  ] [Resource    ] add </usr/share/fonts/truetype/dejavu> in path list
[DEBUG  ] [Resource    ] add </usr/share/fonts/franklin> in path list
[DEBUG  ] [Resource    ] add </usr/local/share/fonts> in path list
[DEBUG  ] [Resource    ] add </usr/local/lib/python3.7/dist-packages/kivy/data/fonts> in path list
[DEBUG  ] [Atlas       ] Load </usr/local/lib/python3.7/dist-packages/kivy/data/images/defaulttheme.atlas>
[DEBUG  ] [Atlas       ] Need to load 1 images
[DEBUG  ] [Atlas       ] Load </usr/local/lib/python3.7/dist-packages/kivy/data/images/defaulttheme-0.png>
[DEBUG  ] [ImageSDL2   ] Load </usr/local/lib/python3.7/dist-packages/kivy/data/images/defaulttheme-0.png>
[DEBUG  ] [ImageSDL2   ] Load </home/myself/progress_track.png>
[DEBUG  ] [ImageSDL2   ] Load </home/myself/progress_bar.png>
[INFO   ] [GL          ] NPOT texture support is available
[INFO   ] [Clipboard   ] Provider: xclip
[INFO   ] [CutBuffer   ] cut buffer support enabled
[DEBUG  ] [Base        ] Create provider from probesysfs
[DEBUG  ] [ProbeSysfs  ] using probesysfs!
[DEBUG  ] [ProbeSysfs  ] found device: Microsoft  Microsoft Basic Optical Mouse v2.0 at /dev/input/event3
[INFO   ] [ProbeSysfs  ] device match: /dev/input/event3
[INFO   ] [MTD         ] Read event from </dev/input/event3>
[DEBUG  ] [ProbeSysfs  ] found device: Chicony HP Elite USB Keyboard System Control at /dev/input/event1
[INFO   ] [ProbeSysfs  ] device match: /dev/input/event1
[INFO   ] [MTD         ] Read event from </dev/input/event1>
[DEBUG  ] [ProbeSysfs  ] found device: Chicony HP Elite USB Keyboard Consumer Control at /dev/input/event2
[INFO   ] [ProbeSysfs  ] device match: /dev/input/event2
[INFO   ] [MTD         ] Read event from </dev/input/event2>
[DEBUG  ] [ProbeSysfs  ] found device: Chicony HP Elite USB Keyboard at /dev/input/event0
[INFO   ] [ProbeSysfs  ] device match: /dev/input/event0
[INFO   ] [MTD         ] Read event from </dev/input/event0>
[INFO   ] [Base        ] Start application main loop
[INFO   ] [MTD         ] </dev/input/event3> range position X is 0 - 0
[INFO   ] [MTD         ] </dev/input/event1> range position X is 0 - 0
[INFO   ] [MTD         ] </dev/input/event3> range position Y is 0 - 0
[INFO   ] [MTD         ] </dev/input/event1> range position Y is 0 - 0
[INFO   ] [MTD         ] </dev/input/event2> range position X is 0 - 0
[INFO   ] [MTD         ] </dev/input/event0> range position X is 0 - 0
[INFO   ] [MTD         ] </dev/input/event3> range touch major is 0 - 0
[INFO   ] [MTD         ] </dev/input/event1> range touch major is 0 - 0
[INFO   ] [MTD         ] </dev/input/event2> range position Y is 0 - 0
[INFO   ] [MTD         ] </dev/input/event0> range position Y is 0 - 0
[INFO   ] [MTD         ] </dev/input/event3> range touch minor is 0 - 0
[INFO   ] [MTD         ] </dev/input/event1> range touch minor is 0 - 0
[INFO   ] [MTD         ] </dev/input/event2> range touch major is 0 - 0
[INFO   ] [MTD         ] </dev/input/event0> range touch major is 0 - 0
[INFO   ] [MTD         ] </dev/input/event3> range pressure is 0 - 255
[INFO   ] [MTD         ] </dev/input/event1> range pressure is 0 - 255
[INFO   ] [MTD         ] </dev/input/event2> range touch minor is 0 - 0
[INFO   ] [MTD         ] </dev/input/event0> range touch minor is 0 - 0
[INFO   ] [MTD         ] </dev/input/event3> axes invertion: X is 0, Y is 0
[INFO   ] [MTD         ] </dev/input/event1> axes invertion: X is 0, Y is 0
[INFO   ] [MTD         ] </dev/input/event2> range pressure is 0 - 255
[INFO   ] [MTD         ] </dev/input/event0> range pressure is 0 - 255
[INFO   ] [MTD         ] </dev/input/event3> rotation set to 0
[INFO   ] [MTD         ] </dev/input/event1> rotation set to 0
[INFO   ] [MTD         ] </dev/input/event2> axes invertion: X is 0, Y is 0
[INFO   ] [MTD         ] </dev/input/event0> axes invertion: X is 0, Y is 0
[INFO   ] [MTD         ] </dev/input/event2> rotation set to 0
[INFO   ] [MTD         ] </dev/input/event0> rotation set to 0
````

**Comment by clickworkorange** [comment]
Can anyone perhaps help me help myself by explaining how physical keyboard events are read by Kivy when running under SDL2 (i.e. _not_ under X11)? What component is responsible for this (xkb? sdl2? something else?) and what does Kivy do with these events to (for example) print the characters in a TextInput? Any helpful pointers appreciated!

**Comment by misl6** [comment]
Hi @clickworkorange,

On your RPi, Kivy is running with the `egl_rpi` window provider, as we can see from the logs:
```
[INFO   ] [Window      ] Provider: egl_rpi
```

So, SDL2, pygame or X11 window implementations are not responsible for sending input events, instead `hidinput` is.

The implementation for `hidinput` could be find here: https://github.com/kivy/kivy/blob/master/kivy/input/providers/hidinput.py

We likely have an issue during the events filtering for this provider, but unfortunately, I do not have a setup where I can test it ATM. (But PRs are welcome)

As an alternative, you can build Kivy with SDL2 support and then explicitly ask to have a SDL2-managed Window, via `export KIVY_WINDOW=sdl2`

**Comment by clickworkorange** [comment]
@misl6 Thanks! I'll see what I can do. I take it I would lose hardware acceleration if I switch to SDL2 as window provider? I'm thinking specifically for the shader transitions...

**Comment by clickworkorange** [comment]
@misl6 Initial research suggests using `xkb` to do the key mapping, so that the keyboard layout set in `/etc/default/keyboard` is honoured (`xkb` is already a requirement). Looking at `xkb` Python bindings that can perform this mapping.

**Comment by clickworkorange** [comment]
As a quick test I made a new translation table to substitute the `keyboard_keys` in `hidinput.py` with one that matches my physical keyboard layout - but then I found there's another layer of translation in the base `Keyboard` class (in [kivy/core/window/\_\_init\_\_.py](https://github.com/kivy/kivy/blob/008c6aed4985a0c272a40e6b74e91141fd9c5074/kivy/core/window/__init__.py#L64)) with a lookup table called `keycodes`. Any characters not found in the table are rejected, and since this table is hard-coded in the base class of the `Window` it appears to prevent the use of any non-English characters, regardless of the provider. I am unsure how I'm supposed to deal with this.

**Comment by clickworkorange** [comment]
On the assumption that this table was mostly just a unicode table I added the characters needed for the language used by my application with their corresponding unicode decimal representations - seems to have worked; I can now type those characters in a TextInput in Kivy.

**Comment by clickworkorange** [comment]
Just wanted to report that I am able to replicate similar behaviour on my laptop running Devuan Daedalus (aka Debian Bookworm), with stock compiled Kivy 2.2.0, SDL 2.26.2, Python 3.11.1, using pygame 2.1.2 as the window provider and hidinput as the keyboard provider (I did not chose to use pygame as the window provider, that's just how it ended up after building and installing). 

Pressing e.g. caps lock produces the text "capslock" in a TextInput, and does not switch case. Oddly I get each keypress repeated twice, so pressing [a] results in "aa" being printed in the TextInput. Even stranger, if I switch the keyboard layout to Swedish in X11/XFCE only the _second_ repeated character is using the localised keymap - e.g. I press [å] and I get "[å" in the Text Input (with the letter "å" sharing the key used for "[" on an English keyboard). That's like... yeah, weird. 

In any case, this means the issue is not entirely rpi_egl dependent and it should be possible to work on a fix without having to run the code on a Raspberry Pi. The only change I made after building master was to edit config.ini to set the keyboard provider to `hidinput,/dev/inputs/event0`. I also had to add myself to the "input" group. Startup log: 

````
[INFO   ] [Logger      ] Record log in /home/myself/.kivy/logs/kivy_23-03-12_16.txt
[INFO   ] [Kivy        ] v2.2.0.dev0
[INFO   ] [Kivy        ] Installed at "/home/myself/Sources/kivy/kivy/__init__.py"
[INFO   ] [Python      ] v3.11.1 (main, Dec 31 2022, 10:23:59) [GCC 12.2.0]
[INFO   ] [Python      ] Interpreter at "/usr/bin/python3"
[INFO   ] [Logger      ] Purge log fired. Processing...
[INFO   ] [Logger      ] Purge finished!
[INFO   ] [Factory     ] 190 symbols loaded
pygame 2.1.2 (SDL 2.26.2, Python 3.11.1)
Hello from the pygame community. https://www.pygame.org/contribute.html
[INFO   ] [Image       ] Providers: img_tex, img_dds, img_pygame, img_pil (img_ffpyplayer ignored)
[INFO   ] [Window      ] Provider: pygame
[WARNING] [Deprecated  ] Pygame has been deprecated and will be removed after 1.11.0: Call to deprecated function __init__ in /home/myself/Sources/kivy/kivy/core/window/window_pygame.py line 42.Called from /home/myself/Sources/kivy/kivy/core/__init__.py line 71 by core_select_lib().
[INFO   ] [GL          ] Using the "OpenGL" graphics system
[INFO   ] [GL          ] Backend used <gl>
[INFO   ] [GL          ] OpenGL version <b'4.2 (Compatibility Profile) Mesa 22.3.3'>
[INFO   ] [GL          ] OpenGL vendor <b'Intel'>
[INFO   ] [GL          ] OpenGL renderer <b'Mesa Intel(R) HD Graphics 4000 (IVB GT2)'>
[INFO   ] [GL          ] OpenGL parsed version: 4, 2
[INFO   ] [GL          ] Shading version <b'4.20'>
[INFO   ] [GL          ] Texture max size <16384>
[INFO   ] [GL          ] Texture max units <16>
[WARNING] [Deprecated  ] Pygame has been deprecated and will be removed after 1.11.0: Call to deprecated function __init__ in /home/myself/Sources/kivy/kivy/core/image/img_pygame.py line 27.Called from /home/myself/Sources/kivy/kivy/core/image/__init__.py line 460 by load().
[INFO   ] [Window      ] virtual keyboard not allowed, single mode, not docked
[INFO   ] [Text        ] Provider: pygame(['text_pango'] ignored)
[INFO   ] [Clipboard   ] Provider: xclip
[INFO   ] [CutBuffer   ] cut buffer support enabled
[WARNING] [Deprecated  ] Pygame has been deprecated and will be removed after 1.11.0: Call to deprecated function __init__ in /home/myself/Sources/kivy/kivy/core/text/text_pygame.py line 36.Called from /home/myself/Sources/kivy/kivy/uix/textinput.py line 2631 by _get_line_options().
[INFO   ] [HIDInput    ] Read event from </dev/input/event0>
[INFO   ] [Base        ] Start application main loop
[INFO   ] [HIDMotionEvent] using <AT Translated Set 2 keyboard>
````

**Comment by clickworkorange** [comment]
Switching to SDL2 as the window provider resolves _all_ issues, while still using hidinput,/dev/input/event0 as the keyboard input. The keyboard behaves perfectly, regardless of selected layout. At least on my laptop. Startup log: 

````
[INFO   ] [Logger      ] Record log in /home/myself/.kivy/logs/kivy_23-03-12_23.txt
[INFO   ] [Kivy        ] v2.2.0.dev0
[INFO   ] [Kivy        ] Installed at "/home/myself/Sources/kivy/kivy/__init__.py"
[INFO   ] [Python      ] v3.11.1 (main, Dec 31 2022, 10:23:59) [GCC 12.2.0]
[INFO   ] [Python      ] Interpreter at "/usr/bin/python3"
[INFO   ] [Logger      ] Purge log fired. Processing...
[INFO   ] [Logger      ] Purge finished!
[INFO   ] [Factory     ] 190 symbols loaded
[INFO   ] [Image       ] Providers: img_tex, img_dds, img_sdl2, img_pil (img_ffpyplayer ignored)
[INFO   ] [Window      ] Provider: sdl2
[INFO   ] [GL          ] Using the "OpenGL" graphics system
[INFO   ] [GL          ] Backend used <sdl2>
[INFO   ] [GL          ] OpenGL version <b'4.2 (Compatibility Profile) Mesa 22.3.3'>
[INFO   ] [GL          ] OpenGL vendor <b'Intel'>
[INFO   ] [GL          ] OpenGL renderer <b'Mesa Intel(R) HD Graphics 4000 (IVB GT2)'>
[INFO   ] [GL          ] OpenGL parsed version: 4, 2
[INFO   ] [GL          ] Shading version <b'4.20'>
[INFO   ] [GL          ] Texture max size <16384>
[INFO   ] [GL          ] Texture max units <16>
[INFO   ] [Window      ] auto add sdl2 input provider
[INFO   ] [Window      ] virtual keyboard not allowed, single mode, not docked
[INFO   ] [Text        ] Provider: sdl2(['text_pango'] ignored)
[INFO   ] [Clipboard   ] Provider: xclip
[INFO   ] [CutBuffer   ] cut buffer support enabled
[INFO   ] [HIDInput    ] Read event from </dev/input/event0>
[INFO   ] [Base        ] Start application main loop
[INFO   ] [HIDMotionEvent] using <AT Translated Set 2 keyboard>
[INFO   ] [GL          ] NPOT texture support is available
````
Since this seems to be a fix (and I know it was suggested by @misl6 earlier) I'm now going to try this on the Pi as well.

**Comment by clickworkorange** [comment]
I rebuilt the version of SDL2 I already had on the Pi with 

````
./configure --disable-pulseaudio --disable-esd --disable-video-opengl --disable-video-wayland --disable-video-directfb --disable-video-x11 --disable-video-vulkan --enable-video-kmsdrm --enable-video-opengles
````
and ran `export KIVY_WINDOW=sdl2` before starting my app - startup log looks like this: 
````
[INFO   ] [Logger      ] Record log in /home/myself/.kivy/logs/kivy_23-03-09_17.txt
[INFO   ] [Kivy        ] v2.1.0
[INFO   ] [Kivy        ] Installed at "/usr/local/lib/python3.7/dist-packages/kivy/__init__.py"
[INFO   ] [Python      ] v3.7.3 (default, Oct 31 2022, 14:04:00) 
[GCC 8.3.0]
[INFO   ] [Python      ] Interpreter at "/usr/bin/python3"
[INFO   ] [Logger      ] Purge log fired. Processing...
[INFO   ] [Logger      ] Purge finished!
[INFO   ] [Factory     ] 189 symbols loaded
[INFO   ] [Image       ] Providers: img_tex, img_dds, img_sdl2, img_pil (img_ffpyplayer ignored)
[INFO   ] [Window      ] Provider: sdl2(['window_egl_rpi'] ignored)
[INFO   ] [GL          ] Using the "OpenGL ES 2" graphics system
[INFO   ] [GL          ] Backend used <sdl2>
[INFO   ] [GL          ] OpenGL version <b'OpenGL ES 2.0'>
[INFO   ] [GL          ] OpenGL vendor <b'Broadcom'>
[INFO   ] [GL          ] OpenGL renderer <b'VideoCore IV HW'>
[INFO   ] [GL          ] OpenGL parsed version: 2, 0
[INFO   ] [GL          ] Shading version <b'OpenGL ES GLSL ES 1.00'>
[INFO   ] [GL          ] Texture max size <2048>
[INFO   ] [GL          ] Texture max units <8>
[INFO   ] [Window      ] auto add sdl2 input provider
[INFO   ] [Window      ] virtual keyboard not allowed, single mode, not docked
[INFO   ] [Text        ] Provider: sdl2
[INFO   ] [Clipboard   ] Provider: xclip
[INFO   ] [CutBuffer   ] cut buffer support enabled
[INFO   ] [HIDInput    ] Read event from </dev/input/event0>
[INFO   ] [Base        ] Start application main loop
[INFO   ] [HIDMotionEvent] using <Chicony HP Elite USB Keyboard>
````
I was happy to see shader performance was just as good as with the rpi_egl window provider. And the caps lock etc keys on the keyboard no longer produce any text - but unfortunately neither do the regular keys! I can see that return, tab, backspace, delete, page up, page down and the arrow keys all work as they should, but I am unable to type _anything_ other than newlines (with the return key). There are no errors or warnings on the console.

**Comment by clickworkorange** [comment]
One peculiar thing is that if I print the events in my `on_key_down` handler they look identical under the egl_rpi and sdl2 window providers - but under sdl2 the characters never make it to the TextInput. So it's not that keyboard events don't fire under sdl2, they just get lost somehow, before they get to the TextInput. 

````
#!/usr/bin/env python3

from kivy.app import App
from kivy.lang import Builder
from kivy.core.window import Window
from kivy.uix.textinput import TextInput

kv = '''
TextInput:
    focus: True
    font_size: dp(64)
'''

class MyApp(App):
  def __init__(self, **kwargs):
    super(MyApp, self).__init__(**kwargs)
    self.keyboard = Window.request_keyboard(self.on_keyboard_close, self)
    self.keyboard.bind(on_key_down=self._on_key_down)

  def _on_key_down(self, keyboard, keycode, text, modifiers):
    print(keycode)
    return False

  def on_keyboard_close(self):
    return True

  def build(self):
    return Builder.load_string(kv)

if __name__ == '__main__':
  app = MyApp()
  try:
    app.run()
  except KeyboardInterrupt:
    app.stop()
````

**Comment by clickworkorange** [comment]
While little more than a crude hack, I have resolved to sticking with `egl_rpi` window manager and changing the `keyboard_keys` dictionary in `providers/hidinput.py` to include the Swedish characters needed for the current project I'm working on. I also remapped all the non-AltGr symbols to match a Swedish keyboard layout: 

```` python
    keyboard_keys = {
        0x01: ('escape', ),
        0x02: ('1', '!'),
        0x03: ('2', '"'),
        0x04: ('3', '#'),
        0x05: ('4', '¤'),
        0x06: ('5', '%'),
        0x07: ('6', '&'),
        0x08: ('7', '/'),
        0x09: ('8', '('),
        0x0a: ('9', ')'),
        0x0b: ('0', '='),
        0x0c: ('+', '?'),
        0x0d: ('´', '`'),
        0x0e: ('backspace', ),
        0x0f: ('tab', ),
        0x10: ('q', 'Q'),
        0x11: ('w', 'W'),
        0x12: ('e', 'E'),
        0x13: ('r', 'R'),
        0x14: ('t', 'T'),
        0x15: ('y', 'Y'),
        0x16: ('u', 'U'),
        0x17: ('i', 'I'),
        0x18: ('o', 'O'),
        0x19: ('p', 'P'),
        0x1a: ('å', 'Å'),
        0x1b: ('¨', '^'),
        0x1c: ('enter', ),
        0x1d: ('lctrl', ),
        0x1e: ('a', 'A'),
        0x1f: ('s', 'S'),
        0x20: ('d', 'D'),
        0x21: ('f', 'F'),
        0x22: ('g', 'G'),
        0x23: ('h', 'H'),
        0x24: ('j', 'J'),
        0x25: ('k', 'K'),
        0x26: ('l', 'L'),
        0x27: ('ö', 'Ö'),
        0x28: ('ä', 'Ä'),
        0x29: ('|', '\\'), # not true
        0x2a: ('shift', ),
        0x2b: ("'", '*'),
        0x2c: ('z', 'Z'),
        0x2d: ('x', 'X'),
        0x2e: ('c', 'C'),
        0x2f: ('v', 'V'),
        0x30: ('b', 'B'),
        0x31: ('n', 'N'),
        0x32: ('m', 'M'),
        0x33: (',', ';'),
        0x34: ('.', ':'),
        0x35: ('-', '_'),
        0x36: ('shift', ),
        0x37: ('numpadmul', ),
        0x38: ('alt', ),
        0x39: ('spacebar', ),
        0x3a: ('capslock', ),
        0x3b: ('f1', ),
        0x3c: ('f2', ),
        0x3d: ('f3', ),
        0x3e: ('f4', ),
        0x3f: ('f5', ),
        0x40: ('f6', ),
        0x41: ('f7', ),
        0x42: ('f8', ),
        0x43: ('f9', ),
        0x44: ('f10', ),
        0x45: ('numlock', ),
        0x46: ('Screenlock', ),
        0x47: ('numpad7', 'home'),
        0x48: ('numpad8', 'up'),
        0x49: ('numpad9', 'pageup'),
        0x4a: ('numpadsubstract', ),
        0x4b: ('numpad4', 'left'),
        0x4c: ('numpad5', ),
        0x4d: ('numpad6', 'right'),
        0x4e: ('numpadadd', ),
        0x4f: ('numpad1', 'end'),
        0x50: ('numpad2', 'down'),
        0x51: ('numpad3', 'pagedown'),
        0x52: ('numpad0', 'insert'),
        0x53: ('numpaddecimal', 'delete'),
        0x54: ('Alt+SysRq', ),
        0x56: ('<', '>'),
        0x57: ('f11', ),
        0x58: ('f12', ),
        0x60: ('numpadenter', ),
        0x61: ('rctrl', ),
        0x62: ('numpaddivide', ),
        0x63: ('print', ),
        0x64: ('alt-gr', ),
        0x66: ('home', ),
        0x67: ('up', ),
        0x68: ('pageup', ),
        0x69: ('left', ),
        0x6a: ('right', ),
        0x6b: ('end', ),
        0x6c: ('down', ),
        0x6d: ('pagedown', ),
        0x6e: ('insert', ),
        0x6f: ('delete', ),
        0x77: ('pause', ),
        0x7D: ('super', ),
        0x7e: ('super', ),
        0x7f: ('compose', ),
        0xff: ('non-US-1', )

        # TODO combinations
        # e0-37    PrtScr
        # e0-46    Ctrl+Break
        # e0-5b    LWin (USB: LGUI)
        # e0-5c    RWin (USB: RGUI)
        # e0-5d    Menu
        # e0-5f    Sleep
        # e0-5e    Power
        # e0-63    Wake
        # e0-38    RAlt
        # e0-1d    RCtrl
        # e0-52    Insert
        # e0-53    Delete
        # e0-47    Home
        # e0-4f    End
        # e0-49    PgUp
        # e0-51    PgDn
        # e0-4b    Left
        # e0-48    Up
        # e0-50    Down
        # e0-4d    Right
        # e0-35    KP-/
        # e0-1c    KP-Enter
        # e1-1d-45 77      Pause
    }
````

To prevent function keys from printing their string representations in my TextInputs I added empties for F1-F12, caps lock, insert, pause, print screen, etc to the `keys_str` dictionary in the same file: 

```` python
    keys_str = {
        'spacebar': ' ',
        'tab': '        ',
        'shift': '',
        'alt': '',
        'ctrl': '',
        'capslock': '',
        'Screenlock': '', # sic
        'numlock': '',
        'compose': '',
        'super': '',
        'escape': '',
        'insert': '',
        'pause': '',
        'print': '',
        'f1': '',
        'f2': '',
        'f3': '',
        'f4': '',
        'f5': '',
        'f6': '',
        'f7': '',
        'f8': '',
        'f9': '',
        'f10': '',
        'f11': '',
        'f12': '',
        'numpad1': '1',
        'numpad2': '2',
        'numpad3': '3',
        'numpad4': '4',
        'numpad5': '5',
        'numpad6': '6',
        'numpad7': '7',
        'numpad8': '8',
        'numpad9': '9',
        'numpad0': '0',
        'numpadmul': '*',
        'numpaddivide': '/',
        'numpadadd': '+',
        'numpaddecimal': '.',
        'numpadsubstract': '-',
    }
````

Finally I've also added the unicode decimal values for the Swedish characters to the `keycodes` dictionary in `window/__init__.py` (not sure what's up with the trailing comma, but I left it as I found it): 

keycodes = {
        # specials keys
        'backspace': 8, 'tab': 9, 'enter': 13, 'rshift': 303, 'shift': 304,
        'alt': 308, 'rctrl': 306, 'lctrl': 305,
        'super': 309, 'alt-gr': 307, 'compose': 311, 'pipe': 310,
        'capslock': 301, 'escape': 27, 'spacebar': 32, 'pageup': 280,
        'pagedown': 281, 'end': 279, 'home': 278, 'left': 276, 'up':
        273, 'right': 275, 'down': 274, 'insert': 277, 'delete': 127,
        'numlock': 300, 'print': 144, 'screenlock': 145, 'pause': 19,

        # a-z keys
        'a': 97, 'b': 98, 'c': 99, 'd': 100, 'e': 101, 'f': 102, 'g': 103,
        'h': 104, 'i': 105, 'j': 106, 'k': 107, 'l': 108, 'm': 109, 'n': 110,
        'o': 111, 'p': 112, 'q': 113, 'r': 114, 's': 115, 't': 116, 'u': 117,
        'v': 118, 'w': 119, 'x': 120, 'y': 121, 'z': 122,

        # 0-9 keys
        '0': 48, '1': 49, '2': 50, '3': 51, '4': 52,
        '5': 53, '6': 54, '7': 55, '8': 56, '9': 57,

        # numpad
        'numpad0': 256, 'numpad1': 257, 'numpad2': 258, 'numpad3': 259,
        'numpad4': 260, 'numpad5': 261, 'numpad6': 262, 'numpad7': 263,
        'numpad8': 264, 'numpad9': 265, 'numpaddecimal': 266,
        'numpaddivide': 267, 'numpadmul': 268, 'numpadsubstract': 269,
        'numpadadd': 270, 'numpadenter': 271,

        # F1-15
        'f1': 282, 'f2': 283, 'f3': 284, 'f4': 285, 'f5': 286, 'f6': 287,
        'f7': 288, 'f8': 289, 'f9': 290, 'f10': 291, 'f11': 292, 'f12': 293,
        'f13': 294, 'f14': 295, 'f15': 296,

        # other keys
        '(': 40, ')': 41,
        '[': 91, ']': 93,
        '{': 123, '}': 125,
        ':': 58, ';': 59,
        '=': 61, '+': 43,
        '-': 45, '_': 95,
        '/': 47, '*': 42,
        '?': 47,
        '`': 96, '~': 126,
        '´': 180, '¦': 166,
        '\\': 92, '|': 124,
        '"': 34, "'": 39,
        ',': 44, '.': 46,
        '<': 60, '>': 62,
        '@': 64, '!': 33,
        '#': 35, '$': 36,
        '%': 37, '^': 94,
        '&': 38, '¬': 172,
        '¨': 168, '…': 8230,
        'ù': 249, 'à': 224,
        'é': 233, 'è': 232,
        'å': 229, 'Å': 197, 
        'ä': 228, 'Ä': 196, 
        'ö': 246, 'Ö': 214, 
    }
````
My local modifications mean, of course, that only a Swedish keyboard will produce the expected characters, but that's all I need at the moment. And my modifications do not provide for caps lock functionality, AltGr symbols (many of which are frequently used on a Swedish keyboard, e.g. "@" and "\"), or keyboard repetition when a key is held down. Good enough for my application though. 

If I was to provide a PR to improve Kivy's keyboard support I would suggest turning the dictionaries I modified into external files, making the choice of which mapping to use configurable through the `config.ini` file. Not a perfect solution by far, but certainly an improvement - and all I have the time, skills and Kivy knowledge to perform well. Let me know if this sounds like something you'd like to see and I'll start working on it.
